sdks/python/apache_beam/runners/worker/sdk_worker_main.py

```diff
@@ -173,22 +173,21 @@ def create_harness(environment, dry_run=False):
   return fn_log_handler, sdk_harness, sdk_pipeline_options
 
 
-def main(unused_argv):
-  """Main entry point for SDK Fn Harness."""
-  (fn_log_handler, sdk_harness,
-   sdk_pipeline_options) = create_harness(os.environ)
+def _start_profiler_if_enabled(sdk_pipeline_options):
   experiments = (sdk_pipeline_options.view_as(DebugOptions).experiments or [])
-  service_name = sdk_pipeline_options.view_as(
-      GoogleCloudOptions).lookup_dataflow_service_option(
-          _ENABLE_GOOGLE_CLOUD_PROFILER, default=os.environ["JOB_NAME"])
-
-  if ((_ENABLE_GOOGLE_CLOUD_PROFILER in experiments) or service_name):
+  gcp_profiler_service_name = sdk_pipeline_options.view_as(
+      GoogleCloudOptions).get_cloud_profiler_service_name(
+          _ENABLE_GOOGLE_CLOUD_PROFILER)
+  if (_ENABLE_GOOGLE_CLOUD_PROFILER in experiments
```

Comment:
I believe there is a behavior change here if the _ENABLE_GOOGLE_CLOUD_PROFILER is specified in experiments without a service name. If I read this code correctly, `gcp_profiler_service_name` will evaluate to `None` now, and we will not enter line 187.

You can catch such errors with better unit tests if you further decouple the logic to identify the cloud profiler configuration, and actually run the cloud profiler. You can introduce the another helper:

```
def _start_profiler(_service_name, _service_version)
```

and write unit tests that will mock out this function, and check what parameters it is being called with (see `unittest.mock.assert_called_with` . Then you can test both the case when cloud profiler is passed as an experiment and when cloud profiler is passed via dataflow service option.

With such (a bit higher level) tests, you don't need to unit test the `get_cloud_profiler_service_name` helper as tests like: 

`test_profiler_uses_provided_service_name_when_specified`
`test_profiler_uses_job_name_when_service_not_specified`
`test_profiler_uses_job_name_when_specified_as_experiment`

will still exercise the  `get_cloud_profiler_service_name` helper.

Reply:
this should also take care of the failing coverage check.

Reply:
Done